### PR TITLE
Better variable naming hygiene in TCO. Fixes #2868

### LINKED ISSUE #2868: TCO codegen has bad hygiene, results in undefined values
If you name an argument to a tco'd function `result`, it clashes with the tco generated name `tco_result`, yielding undefined values.

```
            var go = function (__copy_result) {
                return function (__copy_p) {
                    var __tco_result = __copy_result;
                    var __tco_done = false;
                    var __tco_result;
```

### PR DESCRIPTION
**Pull request by houli**
Hey, first time contributor. Need a little guidance on creating a proper test for this patch but here is a small example of the change in action.

Test PureScript:
```purescript
module Main where

import Prelude

fib :: Int -> Int
fib n = go 0 1 n
  where go result done n = if n > 0
                           then go done (result + done) (n - 1)
                           else result
```

Previous output:
```javascript
"use strict";
var Data_Ord = require("../Data.Ord");
var Data_Ring = require("../Data.Ring");
var Data_Semiring = require("../Data.Semiring");
var Prelude = require("../Prelude");
var fib = function (n) {
    var go = function (__copy_result) {
        return function (__copy_done) {
            return function (__copy_n1) {
                var __tco_result = __copy_result;
                var __tco_done = __copy_done;
                var __tco_done = false;
                var __tco_result;
                function __tco_loop(result, done, n1) {
                    var $0 = n1 > 0;
                    if ($0) {
                        __tco_result = done;
                        __tco_done = result + done | 0;
                        __copy_n1 = n1 - 1 | 0;
                        return;
                    };
                    __tco_done = true;
                    return result;
                };
                while (!__tco_done) {
                    __tco_result = __tco_loop(__tco_result, __tco_done, __copy_n1);
                };
                return __tco_result;
            };
        };
    };
    return go(0)(1)(n);
};
module.exports = {
    fib: fib
};
```

New output:
```javascript
"use strict";
var Data_Ord = require("../Data.Ord");
var Data_Ring = require("../Data.Ring");
var Data_Semiring = require("../Data.Semiring");
var Prelude = require("../Prelude");
var fib = function (n) {
    var go = function (__copy_result) {
        return function (__copy_done) {
            return function (__copy_n1) {
                var __tco_var_result = __copy_result;
                var __tco_var_done = __copy_done;
                var __tco_done = false;
                var __tco_result;
                function __tco_loop(result, done, n1) {
                    var $0 = n1 > 0;
                    if ($0) {
                        __tco_var_result = done;
                        __tco_var_done = result + done | 0;
                        __copy_n1 = n1 - 1 | 0;
                        return;
                    };
                    __tco_done = true;
                    return result;
                };
                while (!__tco_done) {
                    __tco_result = __tco_loop(__tco_var_result, __tco_var_done, __copy_n1);
                };
                return __tco_result;
            };
        };
    };
    return go(0)(1)(n);
};
module.exports = {
    fib: fib
};
```

### CODE REVIEW

**Comment by paf31** [comment]
Looks good to me! @natefaubion, any thoughts?

Can you please update `contributors.md` too?

**Comment by natefaubion** [comment]
Bit of an edge case, but

```purescript
module Main where

import Prelude
import Control.Monad.Eff.Console

__tco_done :: Int
__tco_done = 42

go :: Int -> Int
go n | n < __tco_done = go (n + 1)
     | otherwise = __tco_done
     
main = logShow (go 10)
```

Logs `true`. The previous suggestion to use `$` in the name would fix it I think since it isn't a legal identifier character in PS.

**Comment by houli** [comment]
Ok updated the PR there to use $ as suggested. Fixes @natefaubion's example. Not sure if the same has to be applied to the `copy` variables as they're only ever function arguments?

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ If you would prefer to use different terms, please use the section below instead
 | [@FrigoEU](https://github.com/FrigoEU) | Simon Van Casteren | [MIT license](http://opensource.org/licenses/MIT) |
 | [@garyb](https://github.com/garyb) | Gary Burgess | [MIT license](http://opensource.org/licenses/MIT) |
 | [@hdgarrood](https://github.com/hdgarrood) | Harry Garrood | [MIT license](http://opensource.org/licenses/MIT) |
+| [@houli](https://github.com/houli) | Eoin Houlihan | [MIT license](http://opensource.oth/licenses/MIT) |
```
Comment (paf31) [inline CONTRIBUTORS.md]: Your link seems to be broken.

**Comment by paf31** [comment]
Thanks! @natefaubion Good to merge?

**Comment by houli** [comment]
Yep, definitely is. Silly typo

**Pull request merged**